### PR TITLE
Add guidelines for naming packages

### PR DESCRIPTION
**Pull request by jiahao**
Ref: https://groups.google.com/forum/#!topic/julia-dev/Qw1eaMUKYas

Guidelines courtesy of @johnmyleswhite and @joehuchette

### CODE REVIEW

File: doc/manual/packages.rst [hunk 1]
```diff
@@ -309,6 +309,47 @@ Once you do this, the package manager knows your GitHub user name and can config
...
+
+2. Package names should be plural, especially if you want to define a type in
+   your package with a similar name.
+
+  * ``DataFrames`` provides the ``DataFrame`` type.
+  * ``BloomFilters`` provides the ``BloomFilter`` type.
+
+3. Err on the side of clarity, even if clarity seems long-winded to you.
+
+  * If you're aiming to write the definitive Julia tool for data visualization,
```
Comment (mlubin) [inline doc/manual/packages.rst]: The reason for the name change from CVX to Convex was on request from the author of the CVX matlab package, not for increased clarity. I'd also claim that Convex.jl is one particular way to model convex optimization and isn't the "unique source of truth", so I'm not sure it's a good example.

**Comment by mpastell** [comment]
About 2. : Why would the package names need to be plural in general? I think a lot of the names currently aren't.

**Comment by kmsquire** [comment]
@mpastell, it's not that they need to be, but currently (and into the foreseeable future), a package cannot contain a type with the same name as the package.  This is because the name of the module itself is included as a constant symbol within the package namespace.

```julia
julia> module MyTest
       end

julia> names(MyTest)
1-element Array{Symbol,1}:
 :MyTest

julia> module MyTest
       type MyTest end
       end
Warning: replacing module MyTest
ERROR: invalid redefinition of constant MyTest
```
So if you want to create a type with the name `SuperFilter`, an easy way to name the package similarly is to pluralize it to `SuperFilters`.  (John gave the examples of `DataFrames.DataFrame` and `BloomFilters.BloomFilter` above.)

**Comment by tkelman** [comment]
The `if you want to define a type in your package with a similar name` part doesn't apply to all packages though. If your package is named such that you will want to do that, then yes making the package name plural makes perfect sense as a way to accomplish it. I don't think it should be a hard-and-fast rule, I'm trying to find a legalese documents that starts out by defining various "should" "shall" "must" terminology to think if there's a weaker way of recommending plural names.

**Comment by kmsquire** [comment]
@tkelman, agreed.

**Comment by tkelman** [comment]
Maybe just reverse the sentence? `If you want to define a type in your package with a similar name to the package itself, then the package name should be plural.`

**Comment by johnmyleswhite** [comment]
I do think there's something to be said for a consistency principle that keeps most package names plural since that means you don't have to debate when the name shouldn't be plural. Moreover, the phrase `using Foo` in English often sounds ungrammatical when `Foo` is a noun and it's singular. `using Spoons` makes sense; `using Spoon` doesn't.

**Comment by jiahao** [comment]
> Moreover, the phrase `using Foo` in English often sounds ungrammatical when `Foo` is a noun and it's singular.

On one hand we have packages like `DataFrames` that provide the `DataFrame` type, but we also have packages like `JuliaParser` which do not provide a new type, but rather a single canonical `parse` function. Arguably `using JuliaParsers; JuliaParsers.parse(blah)` sounds odder than the singular variety.

**Comment by mlubin** [comment]
IMHO a policy favoring plural names will lead to some bizarre choices. It's fine to state it as a suggestion, but I would give the package authors some leeway to decide if a plural name really makes sense and is necessary on a case by case basis.

**Comment by jiahao** [comment]
I've updated the guidelines taking into account the discussion so far. Hopefully some of the new examples are more sensible.

**Comment by mpastell** [comment]
I thinks thats much better!

**Comment by kmsquire** [comment]
+1

**Comment by jlapeyre** [comment]
The flat package (module) name space will probably get very crowded quickly (are submodules ever distributed separately?). And I suppose it will be common for assumptions about package relations or which application domains will want to use a name to be wrong.  Apparently [perl has 108K modules]( http://learn.perl.org/modules/) . Here is a list of [53K python modules](https://pypi.python.org/pypi).  Will the current scheme of organizing and naming Julia packages have to be changed when there are 10,000 Julia packages ?  The relation between packages/modules/name spaces, etc. in a particular language will influence how packages are organized.

The first sentence of the van Rossum style guide for [naming conventions for Python](https://www.python.org/dev/peps/pep-0008/#naming-conventions) is "The naming conventions of Python's library are a bit of a mess, so we'll never get this completely consistent..." (but this does not address class hierarchy so much)  Here is a proposal for [Naming conventions and recipes related to packaging in python] (https://www.python.org/dev/peps/pep-0423/). 

Perl module file names always encode a name space hierarchy. Some python packages do and some don't--- the idea is to avoid unneeded verbosity. Some people say java class names are overly verbose; others say they planned for big growth.  Is one of these a good model for Julia ?

File: doc/manual/packages.rst [hunk 2]
```diff
@@ -309,6 +309,53 @@ Once you do this, the package manager knows your GitHub user name and can config
...
+
+  * ``BloomFilters`` provides the ``BloomFilter`` type.
+
+  * In contrast, ``JuliaParser`` provides no new type, but instead new
+    functionality in the ``JuliaParser.parse()`` function.
+
+3. Err on the side of clarity, even if clarity seems long-winded to you.
+
+  * ``RandomMatrices`` is a less ambiguous name than ``RndMat`` or ``RMT``,
+    even if the latter are shorter.
```
Comment (stevengj) [inline doc/manual/packages.rst]: "even if" → "even though", since there's no doubt that the latter are shorter.

File: doc/manual/packages.rst [hunk 3]
```diff
@@ -309,6 +309,53 @@ Once you do this, the package manager knows your GitHub user name and can config
...
+
+4. Less systematic names may suit packages that implement one of several
+   possible approaches to its domain.
```
Comment (stevengj) [inline doc/manual/packages.rst]: Number disagreement: "to their domains".  (or "A less systematic name may suit a package that implements...").

**Comment by mlubin** [comment]
No objections here.

**Comment by joehuchette** [comment]
Looks great.

**Comment by StefanKarpinski** [comment]
:100:

**Comment by tkelman** [comment]
backported in 5c065e79c929885debba660f8143e2bee05eeab2

**Pull request merged**